File: 10-edge/05-pipelines-versioning-and-drift/nodejs/features.js
// Layer 10 - Topic 5: the third implementation, in the shape it usually
// arrives in production. (Node.js)
//
// What this demonstrates
//     A small feature computed in the frontend BFF "because it was easier
//     there". Nobody decided to fork the transform; somebody needed one
//     number in a response and forty lines looked cheaper than a service
//     call. Now three implementations exist, each individually reasonable,
//     and they produce three different answers.
//
//     This file's native mode differs from python/features.py's spec in
//     three places, and NONE of them overlap with the three that
//     golang/features.go gets wrong -- which is the actual lesson. The
//     divergences are not a shared misunderstanding you could fix with one
//     memo. They are independent decisions, so they compose:
//
//       window boundary  both ends inclusive, [T-7d, T]. The spec is
//                        half-open. The Go implementation chose (T-7d, T].
//                        Three readings, three answers.
//       rounding         Math.round is half-UP for positive values (and
//                        asymmetric for negatives, which is its own bug
//                        waiting for a refund feature). The spec is
//                        half-to-even; Go's math.Round is half-away-from-
//                        zero. Three languages, three defaults.
//       recency          Math.round of the day difference instead of
//                        floor, so an event 1.6 days old reports as 2.
//
//     Then `--mode conform` implements the written spec exactly and the
//     diff goes to zero -- which is the evidence for the real fix. The
//     transform gets ONE HOME and everything else calls it across a
//     boundary. A third correct copy is not a fix, it is a third thing to
//     keep correct.
//
// What to look for
//     - Run both modes and diff with python3 python/three_way_diff.py.
//     - Note that the Python/Node disagreement count is NOT the same as
//       the Python/Go one, and the Go/Node one is larger than either.
//       Skew compounds; it does not average out.
//
// No dependencies. Reads the same events.csv every implementation reads:
//     node nodejs/features.js
//     node nodejs/features.js --mode conform --out data/features_node_conform.csv

'use strict';

const fs = require('node:fs');
const path = require('node:path');

const DAY_MS = 86_400_000;
const WINDOW_MS = 7 * DAY_MS;
const AS_OF_MS = 1_785_542_400_000; // 2026-08-01T00:00:00Z

function parseArgs(argv) {
  const args = { mode: 'native', in: null, out: null };
  for (let i = 2; i < argv.length; i += 2) {
    const key = argv[i].replace(/^--/, '');
    args[key] = argv[i + 1];
  }
  const root = path.resolve(__dirname, '..');
  args.in ??= path.join(root, 'data', 'events.csv');
  args.out ??= path.join(
    root,
    'data',
    args.mode === 'conform' ? 'features_node_conform.csv' : 'features_node_native.csv',
  );
  return args;
}

/** Round-half-to-even, which JavaScript does not provide. */
function roundHalfEven(value, places) {
  const scale = 10 ** places;
  const scaled = value * scale;
  const floor = Math.floor(scaled);
  const diff = scaled - floor;
  let rounded;
  if (diff > 0.5) rounded = floor + 1;
  else if (diff < 0.5) rounded = floor;
  else rounded = floor % 2 === 0 ? floor : floor + 1;
  return rounded / scale;
}

/** What an unhurried JavaScript developer reaches for. Half-up. */
function roundNative(value, places) {
  const scale = 10 ** places;
  return Math.round(value * scale) / scale;
}

function loadEvents(file) {
  const lines = fs.readFileSync(file, 'utf8').trim().split('\n');
  const events = [];
  for (let i = 1; i < lines.length; i += 1) {
    const [userId, tsMs, amount] = lines[i].split(',');
    events.push({
      userId: Number(userId),
      tsMs: Number(tsMs),
      amount: Number(amount),
    });
  }
  return events;
}

function compute(events, asOf, conform) {
  const byUser = new Map();
  for (const e of events) {
    if (!byUser.has(e.userId)) byUser.set(e.userId, []);
    byUser.get(e.userId).push(e);
  }
  const lower = asOf - WINDOW_MS;
  const ids = [...byUser.keys()].sort((a, b) => a - b);
  const out = [];
  for (const id of ids) {
    let spend = 0;
    let count = 0;
    let latest = -1;
    for (const e of byUser.get(id)) {
      const inWindow = conform
        ? e.tsMs >= lower && e.tsMs < asOf // the spec: half-open
        : e.tsMs >= lower && e.tsMs <= asOf; // "the last seven days, inclusive"
      if (!inWindow) continue;
      spend += e.amount;
      count += 1;
      if (e.tsMs > latest) latest = e.tsMs;
    }
    let avg = 0;
    let recency = -1;
    if (count > 0) {
      const raw = spend / count;
      avg = conform ? roundHalfEven(raw, 2) : roundNative(raw, 2);
      const days = (asOf - latest) / DAY_MS;
      recency = conform ? Math.floor(days) : Math.round(days);
    } else if (!conform) {
      recency = -1; // this one Node happens to get right
    }
    out.push({ userId: id, spend, count, avg, recency });
  }
  return out;
}

function main() {
  const args = parseArgs(process.argv);
  const conform = args.mode === 'conform';

  let events;
  try {
    events = loadEvents(args.in);
  } catch (err) {
    console.error(`cannot read ${args.in}: ${err.message}`);
    console.error('run python3 python/seed_events.py first');
    process.exit(1);
  }

  const rows = compute(events, AS_OF_MS, conform);
  const lines = ['user_id,spend_7d,txn_count_7d,avg_amount_7d,recency_days'];
  for (const r of rows) {
    lines.push(`${r.userId},${r.spend},${r.count},${r.avg.toFixed(2)},${r.recency}`);
  }
  fs.writeFileSync(args.out, `${lines.join('\n')}\n`);

  console.log(`Node.js feature implementation (${args.mode} mode)`);
  console.log(`  input  : ${args.in} (${events.length} events)`);
  console.log(`  output : ${args.out} (${rows.length} users)`);
  if (conform) {
    console.log('\n  This mode implements python/features.py\'s docstring exactly:');
    console.log('  half-open window, round-half-to-even, floor for recency.');
    console.log('  The three-way diff against Python should be zero rows.');
  } else {
    console.log('\n  This mode is the bug, written the way it really happens:');
    console.log('    window    [T-7d, T]  instead of [T-7d, T)');
    console.log('    rounding  Math.round (half up)');
    console.log('    recency   Math.round instead of floor');
    console.log('  Note that none of these are the three Go gets wrong. That is');
    console.log('  the point: independent decisions compose rather than cancel.');
  }
  console.log('\n  Next: python3 python/three_way_diff.py');
}

main();

File: 10-edge/05-pipelines-versioning-and-drift/golang/features.go
// Layer 10 - Topic 5: how training/serving skew is actually born. (Go)
//
// What this demonstrates
//
//	This file is the bug, written on purpose and written the way it
//	really happens. An ingest service needs the same 7-day features the
//	Python pipeline computes. The Python module is not callable from
//	here, and reimplementing forty lines looked cheaper than standing up
//	an RPC. Nobody was careless. Three decisions simply got made again,
//	independently, and two of them came out differently:
//
//	  window boundary  Python's spec is half-open, [T-7d, T). This file
//	                   does (T-7d, T] -- exclusive lower, inclusive
//	                   upper -- which is the other obvious reading of
//	                   "the last seven days" and reads fine in review.
//	  rounding         math.Round is half-away-from-zero. Python's
//	                   round() and the spec are half-to-even. Nothing in
//	                   either language's documentation suggests the
//	                   other one is the default anywhere else.
//	  empty window     returns recency 0, because zero is what a Go
//	                   struct field starts as and no case in this
//	                   function ever assigns -1.
//
//	Then `-mode conform` implements the written spec exactly, and the
//	diff goes to zero. That is the evidence for the actual fix: the
//	transform gets ONE HOME and everything else calls it across a
//	boundary. A second correct copy is not a fix, it is a second thing
//	to keep correct.
//
// What to look for
//   - Run both modes and diff them against the Python output with
//     python3 python/three_way_diff.py. `native` disagrees on a
//     meaningful fraction of users; `conform` on none.
//   - The three causes are separable, and the diff tool attributes each
//     differing row to one of them. "It's a float thing" is not a
//     diagnosis.
//
// No dependencies. Reads the same events.csv every implementation reads:
//
//	cd golang && go run features.go
//	cd golang && go run features.go -mode conform -out ../data/features_go_conform.csv
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
)

const (
	dayMs    = 86_400_000
	windowMs = 7 * dayMs
	asOfMs   = 1_785_542_400_000 // 2026-08-01T00:00:00Z
)

type event struct {
	userID int
	tsMs   int64
	amount int64
}

type features struct {
	userID      int
	spend       int64
	count       int64
	avgCents    float64
	recencyDays int64
}

// roundHalfEven is what the spec asks for, and what Python's round() does.
// It has to be written out here because Go's math package does not ship it:
// math.Round is half-away-from-zero and math.RoundToEven only works on
// whole numbers, not on a chosen number of decimal places.
func roundHalfEven(v float64, places int) float64 {
	scale := math.Pow(10, float64(places))
	return math.RoundToEven(v*scale) / scale
}

// roundNative is what an unhurried Go developer reaches for.
func roundNative(v float64, places int) float64 {
	scale := math.Pow(10, float64(places))
	return math.Round(v*scale) / scale
}

func compute(events []event, asOf int64, conform bool) []features {
	byUser := map[int][]event{}
	for _, e := range events {
		byUser[e.userID] = append(byUser[e.userID], e)
	}
	ids := make([]int, 0, len(byUser))
	for id := range byUser {
		ids = append(ids, id)
	}
	sort.Ints(ids)

	lower := asOf - windowMs
	out := make([]features, 0, len(ids))
	for _, id := range ids {
		var spend, count, latest int64
		latest = -1
		for _, e := range byUser[id] {
			var inWindow bool
			if conform {
				// The spec: lower bound inclusive, upper exclusive.
				inWindow = e.tsMs >= lower && e.tsMs < asOf
			} else {
				// The other obvious reading of "the last seven days".
				inWindow = e.tsMs > lower && e.tsMs <= asOf
			}
			if !inWindow {
				continue
			}
			spend += e.amount
			count++
			if e.tsMs > latest {
				latest = e.tsMs
			}
		}
		f := features{userID: id, spend: spend, count: count}
		if count == 0 {
			f.avgCents = 0
			if conform {
				f.recencyDays = -1 // the spec
			} else {
				f.recencyDays = 0 // the zero value, never assigned otherwise
			}
		} else {
			avg := float64(spend) / float64(count)
			if conform {
				f.avgCents = roundHalfEven(avg, 2)
			} else {
				f.avgCents = roundNative(avg, 2)
			}
			f.recencyDays = (asOf - latest) / dayMs
		}
		out = append(out, f)
	}
	return out
}

func loadEvents(path string) ([]event, error) {
	fh, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer fh.Close()
	r := csv.NewReader(fh)
	rows, err := r.ReadAll()
	if err != nil {
		return nil, err
	}
	out := make([]event, 0, len(rows))
	for i, row := range rows {
		if i == 0 {
			continue // header
		}
		uid, _ := strconv.Atoi(row[0])
		ts, _ := strconv.ParseInt(row[1], 10, 64)
		amt, _ := strconv.ParseInt(row[2], 10, 64)
		out = append(out, event{uid, ts, amt})
	}
	return out, nil
}

func main() {
	in := flag.String("in", "../data/events.csv", "event log")
	out := flag.String("out", "../data/features_go_native.csv", "output")
	mode := flag.String("mode", "native", "native | conform")
	flag.Parse()
	conform := *mode == "conform"

	events, err := loadEvents(*in)
	if err != nil {
		fmt.Fprintf(os.Stderr, "cannot read %s: %v\n", *in, err)
		fmt.Fprintln(os.Stderr, "run python3 python/seed_events.py first")
		os.Exit(1)
	}

	rows := compute(events, asOfMs, conform)

	fh, err := os.Create(*out)
	if err != nil {
		panic(err)
	}
	defer fh.Close()
	w := csv.NewWriter(fh)
	defer w.Flush()
	w.Write([]string{"user_id", "spend_7d", "txn_count_7d", "avg_amount_7d", "recency_days"})
	for _, f := range rows {
		w.Write([]string{
			strconv.Itoa(f.userID),
			strconv.FormatInt(f.spend, 10),
			strconv.FormatInt(f.count, 10),
			strconv.FormatFloat(f.avgCents, 'f', 2, 64),
			strconv.FormatInt(f.recencyDays, 10),
		})
	}

	fmt.Printf("Go feature implementation (%s mode)\n", *mode)
	fmt.Printf("  input  : %s (%d events)\n", *in, len(events))
	fmt.Printf("  output : %s (%d users)\n", *out, len(rows))
	if conform {
		fmt.Println("\n  This mode implements python/features.py's docstring exactly:")
		fmt.Println("  half-open window, round-half-to-even, recency -1 when empty.")
		fmt.Println("  The three-way diff against Python should be zero rows.")
	} else {
		fmt.Println("\n  This mode is the bug, written the way it really happens:")
		fmt.Println("    window        (T-7d, T]  instead of [T-7d, T)")
		fmt.Println("    rounding      math.Round (half away from zero)")
		fmt.Println("    empty window  recency 0 (the zero value) instead of -1")
		fmt.Println("  Every one of those is defensible in isolation, and none of")
		fmt.Println("  them would be caught by a code review of this file alone.")
	}
	fmt.Println("\n  Next: python3 python/three_way_diff.py")
}
